models/Element/Service.php

```diff
@@ -718,31 +718,81 @@ public static function gridElementData(ElementInterface $element)
+        $currentUserId = $user->getId();
+        $userIds = $user->getRoles();
+        $userIds[] = $currentUserId;
+
+        $userWorkspacesSql = '
+            SELECT cpath, cid, list
+            FROM users_workspaces_'.$type.'
+            WHERE userId = '.$currentUserId.'
+        ';
+        $userWorkspaces = $db->fetchAll($userWorkspacesSql);
+
+        //this collects the array that are on user-level, which have top priority
+        $userCid = [];
+        foreach ($userWorkspaces as $userWorkspace){
+            $userCid[] = $userWorkspace['cid'];
         }
 
+        $roleWorkspacesSql = '
+            SELECT
+                cpath, userid, max(list) as list
+            FROM users_workspaces_'.$type.'
+            WHERE userId IN (' . implode(',', $userIds) . ') AND cid NOT IN ('.implode(',', $userCid).')
+            GROUP BY cpath
+        ';
+        $roleWorkspaces = $db->fetchAll($roleWorkspacesSql);
+
+        $allWorkspaces = array_merge($userWorkspaces, $roleWorkspaces);
+
+        $uniquePaths = [];
+        foreach ($allWorkspaces as $workspace){
+            $uniquePaths[$workspace['cpath']] = $workspace['list'];
+        }
+        ksort($uniquePaths);
+
+        //TODO: above this should be all in one query (eg. instead of ksort, use sql sort) but had difficulties making the `group by` working properly to let user permissions take precedence
+
+        $totalPaths = count($uniquePaths);
+
+        $uniquePathsKeys = array_keys($uniquePaths);
+
         $forbidden = [];
-        if (count($workspaces) > 0) {
-            foreach ($workspaces as $workspace) {
-                if (!$workspace->getList()) {
-                    $forbidden[] = $workspace->getCpath();
+        $allowed = [];
+        if ($totalPaths > 0) {
+            for ($index = 0; $index < $totalPaths; $index++) {
+                $path = $uniquePathsKeys[$index];
+                if ($uniquePaths[$path] == 0) {
+                    $forbidden[$path] = [];
+                    for ($findIndex = $index + 1; $findIndex < $totalPaths; $findIndex++) { //NB: the starting index is the last index we got
+                        $findPath = $uniquePathsKeys[$findIndex];
+                        if (str_contains($findPath, $path)) { //it means that we found a children
+                            if ($uniquePaths[$findPath] == 1) {
+                                array_push($forbidden[$path], $findPath); //adding list=1 children
+                            }
+                        } else {
+                            break;
+                        }
+                    }
+                }else{
+                    $allowed[] = $path;
                 }
             }
         } else {
-            $forbidden[] = '/';
+            $forbidden['/'] = [];
         }
 
-        return $forbidden;
+        return array('forbidden' => $forbidden, 'allowed' => $allowed);
```

Comment:
mhh, i've missed that, what about keeping `findForbiddenPaths()` and instead put the new changes from the PR in a new method `getAllowedAndForbiddenPaths()`?

Reply:
since this is internal, I'm fine with fixing the advanced object search ... at some point we also should completely fix the permissions there too. but that is a different task then.